string.c

```diff
@@ -2981,13 +2982,13 @@ str_uplus(VALUE str)
 /*
  * call-seq:
  *   -string -> frozen_string
+ *   string.dedup -> frozen_string
```

Comment:
This change violates the recommendations in the documentation guide (`doc/contributing/documentation_guide.md`), and should not be included.

Reply:
Oh, I wasn't aware about this guide, sorry.
Unfortunately, it seems that when it was merged, documentation rendering by RDoc wasn't considered in introducing those rules. This makes docs strictly worse and less useful.

When call-sequences for two aliases are specified, RDoc are (typically) smart enough to distinguish them, so, in 3.0 (before the new guide was implemented), when one navigated, say, to `Enumerable#collect_compact` [docs](https://docs.ruby-lang.org/en/3.0/Enumerable.html#method-i-collect_concat), they've seen its real call-sequence:
![image](https://user-images.githubusercontent.com/129656/219744985-049ed5d5-7def-44c4-9531-71f56b887d2c.png)
...and can immediately start to use it (even if `flat_map` is used in the examples).

What happens [today](https://docs.ruby-lang.org/en/3.2/Enumerable.html#method-i-collect_concat)? RDoc still tries to render the alias you navigated to, but can't guess the call-sequence, and one have this:
![image](https://user-images.githubusercontent.com/129656/219747083-93ac90ac-bb5d-4dbc-be59-768339dadade.png)

With `String#dedup` it is even worse. The distinction of `-string` and `string.dedup` is important enough to be discussed for half a year, and RDoc, unfortunately, is absolutely helpless here (maybe due to weird processing of `-foo`, need to debug), and [gives this](https://docs.ruby-lang.org/en/3.2/String.html#method-i-dedup):

![image](https://user-images.githubusercontent.com/129656/219747893-97afa616-a2ab-4236-9b77-ecbdc2fee9c0.png)

...so one navigates to `#dedup` from the String TOC, and they see `-string` as the only possible call-sequence. (Note also another side effect of the 2021 guide: it has both auto-rendering of "alias for @-" (which is to the point and contextual, if you navigate to `-@`, it will say "Also aliased as: dedup") and vague hand-written "String#dedup is an alias for String#-@."

So, I believe what should be done is:
* Merge this with aliases;
* Change the guide back to suggest to **always include aliases** in call-seq (and stop suggesting writing about them in free-form text);
* Fix back all the core docs that were "improved" since the guide introduction.

Reply:
> Oh, I wasn't aware about this guide, sorry. Unfortunately, it seems that when it was merged, documentation rendering by RDoc wasn't considered in introducing those rules. This makes docs strictly worse and less useful.
> 
> When call-sequences for two aliases are specified, RDoc are (typically) smart enough to distinguish them, so, in 3.0 (before the new guide was implemented),  when one navigated, say, to `Enumerable#collect_compact` [docs](https://docs.ruby-lang.org/en/3.0/Enumerable.html#method-i-collect_concat), they've seen its real call-sequence: ![image](https://user-images.githubusercontent.com/129656/219744985-049ed5d5-7def-44c4-9531-71f56b887d2c.png) ...and can immediately start to use it (even if `flat_map` is used in the examples).

The guide was committed August 2020, before 3.0 (it eventually was moved and renamed).  However, even in older versions, the call-seq filtering appears to be present.  Having call-seq documentation mention the alias but having all examples in the documentation use the canonical method is confusing and something we should avoid, IMO.

> What happens [today](https://docs.ruby-lang.org/en/3.2/Enumerable.html#method-i-collect_concat)? RDoc still tries to render the alias you navigated to, but can't guess the call-sequence, and one have this: ![image](https://user-images.githubusercontent.com/129656/219747083-93ac90ac-bb5d-4dbc-be59-768339dadade.png)

Admittedly, this is not good.  My preference would be to have RDoc detect when the method is an alias and the call-seq does not mention it, only link to the alias used in the call-seq, and not show a call-seq or other description.  I think we should have the documentation nudge people in the direction of using the canonical method, instead of trying to treat each alias equally and duplicating the documentation for each alias.  This avoids the confusion when the call-seq lists the alias, and all example code and explanatory text uses the canonical method.  It also avoids redundancy.

> With `String#dedup` it is even worse. The distinction of `-string` and `string.dedup` is important enough to be discussed for half a year, and RDoc, unfortunately, is absolutely helpless here (maybe due to weird processing of `-foo`, need to debug), and [gives this](https://docs.ruby-lang.org/en/3.2/String.html#method-i-dedup):
> 
> ![image](https://user-images.githubusercontent.com/129656/219747893-97afa616-a2ab-4236-9b77-ecbdc2fee9c0.png)
> 
> ...so one navigates to `#dedup` from the String TOC, and they see `-string` as the only possible call-sequence. (Note also another side effect of the 2021 guide: it has both auto-rendering of "alias for @-" (which is to the point and contextual, if you navigate to `-@`, it will say "Also aliased as: dedup") and vague hand-written "String#dedup is an alias for String#-@."

The alias instructions in the guide were committed before RDoc listed aliases automatically (4a5dd970a750e90a9fe98d66babfa53665429221). See https://docs.ruby-lang.org/en/2.7.0/Hash.html#method-i-key-3F for how it looked in Ruby 2.7.  I agree now it no longer makes sense to mention aliases, since RDoc now includes them.

> So, I believe what should be done is:
> 
> * Merge this with aliases;
> * Change the guide back to suggest to **always include aliases** in call-seq (and stop suggesting writing about them in free-form text);
> * Fix back all the core docs that were "improved" since the guide introduction.

I'm not sure what you mean by "Merge this with aliases".  I'm against including aliases in call-seq.  I think we should take the following steps:

* Change the documentation guide to not mention listing aliases manually, since RDoc starting in 3.0 now takes care of it automatically. 
* Fix the cases where aliases are listed manually.
* Update RDoc to have aliases link to the method shown in call-seq, and not repeat documentation for aliases.

Reply:
> The guide was committed August 2020, before 3.0 (it eventually was moved and renamed). However, even in older versions, the call-seq filtering appears to be present. 

If by call-seq filtering you mean "gradual removing of alias call-seq", the explanation is simple: the guide was committed by Burdette Lamar and represents his understanding of the matter. He is also gradually rewriting docs for all the core classes, some before 3.0, some after, following the same understanding. 

Enumerable lost aliases after 3.0 ([3.0](https://docs.ruby-lang.org/en/3.0/Enumerable.html#method-i-collect_concat), [3.1](https://docs.ruby-lang.org/en/3.1/Enumerable.html#method-i-collect_concat)), but, say, Hash, after 2.7 ([2.7](https://docs.ruby-lang.org/en/2.7/Hash.html#method-i-filter), [3.0](https://docs.ruby-lang.org/en/3.0/Hash.html#method-i-filter), was rendered quite weird back then).

> Having call-seq documentation mention the alias but having all examples in the documentation use the canonical method is confusing and something we should avoid, IMO.

Well, it is at least better than having the alias, but not a call-sequence for it :) When you navigate to `collect_compact`, and it at least has a clear signature description and mentions that it is an alias for `filter_map`, the examples with `filter_map` is somewhat justified.
I think it might've been better if `Alias for: #filter_map` would immediately follow the signature, then it would be read naturally: "this is method Y, which is an alias for method X, here are examples with the original method X, they'll work with Y obviously."

> My preference would be to have RDoc detect when the method is an alias and the call-seq does not mention it, only link to the alias used in the call-seq, and not show a call-seq or other description.

That might work too (but see the clarification below), but it requires changes in RDoc to be made, while the docs became sub-par now (and can be returned to better form with a relatively low effort).

>  I think we should have the documentation nudge people in the direction of using the canonical method, instead of trying to treat each alias equally and duplicating the documentation for each alias.

This might be a good direction, but there is a nuance here: the "canonical" method, according to the method definition code, is not always the one that became preferred by the community. To name a few:
* `yield_self` introduced earlier but has very unfortunate name, with `then` being clearly superior but just an alias;
* `inject` is an original method (inherited, as far as I can guess, from Smalltalk), while `reduce` is the industry standard and preferred in most codebases I saw;
* same with `collect` vs. `map`
* `filter` was introduced as a more modern alias for `select` (aligned with industry), but which is "preferred"?..
* `Hash#each_pair` is canonical, but every codebase I saw uses `Hash#each` (which is an alias)

Then, the situation with this particular method (`-`/`dedup`) is even more complicated: neither is "preferred", and that was the point of aliasing: in some situations, `-"foo"` is more expressive and short, in others, it is not convenient. So, would I provide a code example for this method (which I should've, but was too lazy!), I'd probably do it this way:
```ruby
FILENAME = -'test.txt' # gives deduplicated string

# In method chains, it is sometimes more convenient to use #dedup form
host.append(path).dedup.then { known_urls << _1 }
# or 
make_urls.map(&:dedup)
```

> The alias instructions in the guide were committed before RDoc listed aliases automatically

Yeah, I came to the same conclusion eventually :)

> I'm against including aliases in call-seq. 

Can you justify this, please? (Given that "let's fix RDoc right now" is off the table... unless it isn't?.. But then, how exactly should it be fixed?)

1. In general, right now, we have a lot of docs that are worse than they've been (like very core and classic [Enumerable#reduce](https://docs.ruby-lang.org/en/3.2/Enumerable.html#method-i-reduce)). It is relatively easy to fix by returning back alias call-sequences (I have PR almost ready). Having the choice between "leave it as it is" vs. "at least make it have clear call-sequences for base methods," I don't see why to choose the former.
2. But in particular, for `String#-` and `String#dedup`, as I argue above, _even_ if RDoc would be changed to "we have only one entry of docs, and the rest just says 'see the other method'", at least some mention of two possible convenience call-sequences should be in the docs.

Reply:
> > The guide was committed August 2020, before 3.0 (it eventually was moved and renamed). However, even in older versions, the call-seq filtering appears to be present.
> 
> If by call-seq filtering you mean "gradual removing of alias call-seq", the explanation is simple: the guide was committed by Burdette Lamar and represents his understanding of the matter. He is also gradually rewriting docs for all the core classes, some before 3.0, some after, following the same understanding.

No, the guide was originally written and committed by me (d43e99b7220ae3a7fbd58fbc8bde74085f6bd14c).  Changes to it were based on discussions mostly in GitHub.  For example, how to handle aliases was discussed in #3468.

> Enumerable lost aliases after 3.0 ([3.0](https://docs.ruby-lang.org/en/3.0/Enumerable.html#method-i-collect_concat), [3.1](https://docs.ruby-lang.org/en/3.1/Enumerable.html#method-i-collect_concat)), but, say, Hash, after 2.7 ([2.7](https://docs.ruby-lang.org/en/2.7/Hash.html#method-i-filter), [3.0](https://docs.ruby-lang.org/en/3.0/Hash.html#method-i-filter), was rendered quite weird back then).
> 
> > Having call-seq documentation mention the alias but having all examples in the documentation use the canonical method is confusing and something we should avoid, IMO.
> 
> Well, it is at least better than having the alias, but not a call-sequence for it :) When you navigate to `collect_compact`, and it at least has a clear signature description and mentions that it is an alias for `filter_map`, the examples with `filter_map` is somewhat justified. I think it might've been better if `Alias for: #filter_map` would immediately follow the signature, then it would be read naturally: "this is method Y, which is an alias for method X, here are examples with the original method X, they'll work with Y obviously."

Again, this can be fixed by fixing RDoc to only generate the alias in such cases.

> > My preference would be to have RDoc detect when the method is an alias and the call-seq does not mention it, only link to the alias used in the call-seq, and not show a call-seq or other description.
> 
> That might work too (but see the clarification below), but it requires changes in RDoc to be made, while the docs became sub-par now (and can be returned to better form with a relatively low effort).

I think we have sufficient time before the next release to make the change in RDoc.  We should not make changes now to the documentation that we'll have to revert after fixing RDoc.

> > I think we should have the documentation nudge people in the direction of using the canonical method, instead of trying to treat each alias equally and duplicating the documentation for each alias.
> 
> This might be a good direction, but there is a nuance here: the "canonical" method, according to the method definition code, is not always the one that became preferred by the community. To name a few:
> 
> * `yield_self` introduced earlier but has very unfortunate name, with `then` being clearly superior but just an alias;

FWIW, These are not currently aliases, but separate method definitions in `kernel.rb`.

> * `inject` is an original method (inherited, as far as I can guess, from Smalltalk), while `reduce` is the industry standard and preferred in most codebases I saw;

This is more of a toss-up.  `inject ruby` has about 1.4M hits on GitHub, `reduce ruby` has about 2.2M.  FWIW, I have no problems changing the `inject` documentation from `inject` to `reduce`, even though I typically use `inject` over `reduce`.

> * same with `collect` vs. `map`

`map` is already the only thing listed in call-seq.

> * `filter` was introduced as a more modern alias for `select` (aligned with industry), but which is "preferred"?..

This is another toss up. `select ruby` has about 8.8M hits on GitHub, and `filter` has 6.6M.  Maybe that's because `filter` was only added recently.  Again, not opposed to switching the documentation to use `filter`.

> * `Hash#each_pair` is canonical, but every codebase I saw uses `Hash#each` (which is an alias)

Currently, both are listed in call-seq.  I agree that we should only use `each` since that is much more common.

These and similar issues are fairly simple to fix.

> Then, the situation with this particular method (`-`/`dedup`) is even more complicated: neither is "preferred", and that was the point of aliasing: in some situations, `-"foo"` is more expressive and short, in others, it is not convenient. So, would I provide a code example for this method (which I should've, but was too lazy!), I'd probably do it this way:
> 
> ```ruby
> FILENAME = -'test.txt' # gives deduplicated string
> 
> # In method chains, it is sometimes more convenient to use #dedup form
> host.append(path).dedup.then { known_urls << _1 }
> # or 
> make_urls.map(&:dedup)
> ```

I agree that this is a case where we could make an exception to what the guide currently states.

> > I'm against including aliases in call-seq.
> 
> Can you justify this, please? (Given that "let's fix RDoc right now" is off the table... unless it isn't?.. But then, how exactly should it be fixed?)

I don't think that "Let's fix RDoc before next release" is off the table. RDoc currently scans the call-seq to filter for the alias.  If nothing matches, and the only matching values are for an alias that RDoc is aware of, RDoc can just link to the alias and not display documentation.

> 1. In general, right now, we have a lot of docs that are worse than they've been (like very core and classic [Enumerable#reduce](https://docs.ruby-lang.org/en/3.2/Enumerable.html#method-i-reduce)). It is relatively easy to fix by returning back alias call-sequences (I have PR almost ready). Having the choice between "leave it as it is" vs. "at least make it have clear call-sequences for base methods," I don't see why to choose the former.

Other than this alias issue, can you explain which docs are worse than they've been?  You may be unaware, but a blanket statement like this can be considered insulting to @BurdetteLamar and others who have worked on documentation improvements.

Please don't put forward a false dichotomy. The choice isn't between "leave it as it is" and "add back alias call-seqs".  I've already described how I think this should be handled by fixing RDoc.

> 2. But in particular, for `String#-` and `String#dedup`, as I argue above, _even_ if RDoc would be changed to "we have only one entry of docs, and the rest just says 'see the other method'", at least some mention of two possible convenience call-sequences should be in the docs.

I would be OK update the guide to mention some special cases where including aliases in call-seqs is acceptable, such as one method is a unary operator (`String#-@`, not `String#-`).

Reply:
>> If by call-seq filtering you mean "gradual removing of alias call-seq", the explanation is simple: the guide was committed by Burdette Lamar and represents his understanding of the matter. 
>
> [...] Changes to it were based on discussions mostly in GitHub. For example, how to handle aliases was discussed in https://github.com/ruby/ruby/pull/3468.

I see that the author of the changes of the aliases is indeed Burdette Lamar. Who started to remove aliases by himself even before that. Which is what I was saying. Though I don't actually see that as an important thing to clarify "how it is started" :) What to do now is more important.

>> `inject` is an original method (inherited, as far as I can guess, from Smalltalk), while `reduce` is the industry standard and preferred in most codebases I saw;
> This is more of a toss-up. `inject ruby` has about 1.4M hits on GitHub, `reduce ruby` has about 2.2M. FWIW, I have no problems changing the `inject` documentation from `inject` to `reduce`, even though I typically use `inject` over `reduce`.

The point I was trying to make—which you seem to confirm, not refute—is that changing everything to "there is one main method, the rest is just references/aliases" might require not only changing RDoc, but also doing some decisions of "what's the main method," which might be non-trivial and even unsolvable (e.g., I believe that you are right, and amount of Rubyists that prefer `inject` and those preferring `reduce` would be shared 40/60).

>> same with `collect` vs. `map`

> `map` is already the only thing listed in call-seq.

But it is considered as an alias internally (while "collect" is considered the original method).

I was trying to say that if the change to RDoc _would_ be made before 3.3, it would require settling a lot of such contradictions (e.g., changing Ruby code slightly in many places).

>> In general, right now, we have a lot of docs that are worse than they've been (like very core and classic [Enumerable#reduce](https://docs.ruby-lang.org/en/3.2/Enumerable.html#method-i-reduce)).

> Other than this alias issue, can you explain which docs are worse than they've been?

The aliasing issue is the **only one** I am talking about in this conversation. Only the fact that a lot of methods (including "preferred" or "equally important" aliases) lost their call-sequences. 

Nothing else is the point of this particular discussion. 

Please don't try to antagonize me, I understand that I am not the easiest person to deal with, but I am looking for a compromise and have the utmost respect for everybody working on the documentation.

> Please don't put forward a false dichotomy. The choice isn't between "leave it as it is" and "add back alias call-seqs". I've already described how I think this should be handled by fixing RDoc.

I don't really understand why you are trying to reduce my point. I clarified that everything that I am saying is relevant "if fixing RDoc is off the table."

My point was that there is no public plan to fix RDoc right now and I am not sure whether it is hard or easy, so _for now_ returning call-seq aliases would be just a safety measure in case it wouldn't be done before 3.3. Returning aliases back is half an hour of work. Removing them after the RDoc fix is another half an hour. 

Leaving it as is till RDoc is fixed is an option, though. As we already spent on this discussion more resource than spent on the docs adjustment, I see this option as favorable now :)

> I agree that this is a case where we could make an exception to what the guide currently states.

So, in this particular PR, keeping `dedup` call-sequence is OK? :)

Reply:
> > > `inject` is an original method (inherited, as far as I can guess, from Smalltalk), while `reduce` is the industry standard and preferred in most codebases I saw;
> > > This is more of a toss-up. `inject ruby` has about 1.4M hits on GitHub, `reduce ruby` has about 2.2M. FWIW, I have no problems changing the `inject` documentation from `inject` to `reduce`, even though I typically use `inject` over `reduce`.
> 
> The point I was trying to make—which you seem to confirm, not refute—is that changing everything to "there is one main method, the rest is just references/aliases" might require not only changing RDoc, but also doing some decisions of "what's the main method," which might be non-trivial and even unsolvable (e.g., I believe that you are right, and amount of Rubyists that prefer `inject` and those preferring `reduce` would be shared 40/60).

In most cases, the decision of which method isn't hard because one is obviously more popular.  In the cases where it is a toss-up (closer to even), it doesn't really matter which one is chosen.

> > > same with `collect` vs. `map`
> 
> > `map` is already the only thing listed in call-seq.
> 
> But it is considered as an alias internally (while "collect" is considered the original method).

Incorrect, neither internally is an alias, they are separate methods pointing to the same C function:

```c
rb_define_method(rb_mEnumerable, "collect", enum_collect, 0);
rb_define_method(rb_mEnumerable, "map", enum_collect, 0);
```

> I was trying to say that if the change to RDoc _would_ be made before 3.3, it would require settling a lot of such contradictions (e.g., changing Ruby code slightly in many places).

In the places where the decision has already been made (only primary method is listed in call-seq), all that needs to be done is to remove the explicit listing of aliases, which doesn't need to happen immediately.  If the explicit listing is not removed, the documentation is only slightly redundant.

In the places where the decision has not already been made (documentation has not been updated yet to list only primary method in call-seq), no changes are needed immediately, they can be added later as the documentation is updated to conform with the documentation guide.

The only really important change before 3.3 is fixing RDoc.  Considering I'm the one advocating for that, I'll submit a pull request for it.

> I don't really understand why you are trying to reduce my point. I clarified that everything that I am saying is relevant "if fixing RDoc is off the table."

I think fixing RDoc is very much "on the table".  

> > I agree that this is a case where we could make an exception to what the guide currently states.
> 
> So, in this particular PR, keeping `dedup` call-sequence is OK? :)

Yes, but only if the documentation lists when it is best to use each method.